### DISCUSSION TITLE
Is Remote sampling works without Jaeger-agent sampling port 5778?

### DISCUSSION THREAD

**Discussion by Narenderbhcu** [jaegertracing/jaeger#4142] [Q&A]
### What happened?

Hi, We would like to know how remote sampling works without jaeger-agent sampling report?
we have configured remote sampling using OTEL GO SDK as below(intstead of jaeger-agent port 5778 we have used opentelemetry collector ports 4317 and 4318 but which is not working. 

as per below link, both Jaeger Agent and OpenTelemetry Collector implement the Jaeger sampling service endpoint.

[https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/samplers/jaegerremote]

jaegerRemoteSampler := jaegerremote.New(
"your-service-name",
jaegerremote.WithSamplingServerURL("http://{sampling_service_host_name}:5778"),
jaegerremote.WithSamplingRefreshInterval(10*time.Second),
jaegerremote.WithInitialSampler(trace.TraceIDRatioBased(0.5)),
)

tp := trace.NewTracerProvider(
	trace.WithSampler(jaegerRemoteSampler),
	...
)
otel.SetTracerProvider(tp)

I am trying to use opentelemetry collector sampling port instead of jaeger agent port "http://{sampling_service_host_name}:5778", but I am unable to fetch sampling strategy information from the backend.
Tried with
jaegerremote.WithSamplingServerURL("http://{sampling_service_host_name}:4317"), and jaegerremote.WithSamplingServerURL("http://{sampling_service_host_name}:4318"),

but not working, please suggest your opinion.

### Steps to reproduce

Try to generate traces using otel sdk by applying remote sampling as below and you can see that remote sampling not working as expected.
https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/samplers/jaegerremote
jaegerRemoteSampler := jaegerremote.New(
"your-service-name",
**jaegerremote.WithSamplingServerURL("http://{sampling_service_host_name}:4317"),**
**or**
**jaegerremote.WithSamplingServerURL("http://{sampling_service_host_name}:4318"),**
jaegerremote.WithSamplingRefreshInterval(10*time.Second),
jaegerremote.WithInitialSampler(trace.TraceIDRatioBased(0.5)),
)

### Expected behavior

Jaeger should support remote sampling without Jaeger-agent sampling port 5778.

### Jaeger backend version

v1.38.0

### SDK

Opentelemetry Go SDK

### Pipeline

OTEL SDK -> OTEL Collector -> Jaeger Collector -> Elasticsearch -> Jaeger UI

### Stogage backend

Elasticsearch

### Operating system

Linux

### Deployment model

kubenetes

**Comment by yurishkuro** [comment]
I am not sure about OTEL collector, are you sure it implements the sampling strategy endpoint on the same port as the port used to accept trace data?

With Jaeger, you do not need the agent to serve sampling strategies, Jaeger collector exposes an HTTP endpoint `/api/sampling` on port 14268.

https://www.jaegertracing.io/docs/1.41/deployment/#collector

**Reply by Narenderbhcu** [reply]
We are not sure, it seems like there is no dedicated port for sampling endpoint for opentelemetry collector, If you have any idea on this kindly share.

And thanks, good to know that without agent sampling strategy now works with 14268 port with /api/sampling endpoint. we will uplift to  Jaeger1.41 to support this feature. Thanks.

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro , 
Thanks for the support!
We uplifted Jaeger to 1.41 and while testing observed that, when we disable TLS for collector port 14268, then only we are able to fetch sampling information from Jaeger backend, if TLS is enabled for collector port 14268, its throwing message, as "Client sent an HTTP request to an HTTPS server". how can we fix this? for Jaeger agent 5778 port, did not observe this kind of issue. please provide your opinion.

**Reply by yurishkuro** [reply]
> "Client sent an HTTP request to an HTTPS server"

Sounds like you need to change SDK configuration to https endpoint, to match the server configuration

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro , "Jaeger collector exposes an HTTP endpoint /api/sampling on port 14268"==> I could see this information updated in Jaeger 1.41 documentation but functionality is available in old versions also, so please confirm do we really need to uplift to 1.41to get this feature?

**Reply by Narenderbhcu** [reply]
> > "Client sent an HTTP request to an HTTPS server"
> 
> Sounds like you need to change SDK configuration to https endpoint, to match the server configuration

Tried SDK configuration as below, but not working, it seems it is expecting client certificates,not sure how we can fetch. if you have any reference pleas share.

jaegerRemoteSampler := jaegerremote.New(
		"foo",
		jaegerremote.WithSamplingServerURL("https://collector_host_name:14268/api/sampling"),
		jaegerremote.WithSamplingRefreshInterval(10*time.Second), // decrease polling interval to get quicker feedback
		jaegerremote.WithInitialSampler(trace.TraceIDRatioBased(0.5)),
	)

**Reply by yurishkuro** [reply]
> do we really need to uplift to 1.41to get this feature?

Not, it's been like that for a while, since Dec 2019 (#1990)

**Reply by yurishkuro** [reply]
> Tried SDK configuration as below, but not working, it seems it is expecting client certificates

Client certificates are only useful if you want to authenticate your clients, it's not required for basic TLS.

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro, 

Thanks now with TLS enabled also we are able to fetch the data using jaegerremote.WithSamplingServerURL("https://collector_host_name:14268/api/sampling"),

But now in without agent case, though we are able to fetch the sampling info, but its not getting applied(verified result in Jaeger UI).
When we compare the fetched output of with agent and without agent observed some difference, so asking below question.

For jaeger-agent we set server URL as below and it is working fine able to fetch the data( please note /sampling endpoint did not pass)
jaegerremote.WithSamplingServerURL("https://agent_host_name:5778"),

For without agent case do we need to pass the endpoint /api/sampling or not?

jaegerremote.WithSamplingServerURL("https://collector_host_name:14268), 
(or)
jaegerremote.WithSamplingServerURL("https://collector_host_name:14268/api/sampling"),

Please provide your opinion.

**Reply by yurishkuro** [reply]
you need to pass the full URL: `https://collector_host_name:14268/api/sampling`

**Reply by Narenderbhcu** [reply]
Just attached two files, one file is the output of fetched sampling information of Without Agent and another one withAgent. And the noticed differences in the files are, withAgent case everything working fine, samplingRate is 0.8 at starting and at the end. Without Agent case Sampling Rate 0.001( why this default one came) at the starting and at the end SamplingRate is 0.8. what I need to do to get 0.8 at the starting also in without Agent case. please suggest.
[WithoutAgent_Fetched_sampling_information.log](https://github.com/jaegertracing/jaeger/files/10399107/WithoutAgent_Fetched_sampling_information.log)
[wihtAgentFetchedSampingInfo.log](https://github.com/jaegertracing/jaeger/files/10399108/wihtAgentFetchedSampingInfo.log)

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro , could you please check the provided logs and share your opinion.

**Reply by yurishkuro** [reply]
Looks like it's all working. The initial 0.001 rate is used before the sampler has a chance to retrieve the strategy from the backend for the first time.

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro 
There is a problem, I have tested for 10min, for ervery minute client code will fetch sampling information from backend, the last 10th iteration also i could se samplingRate as 0.001, like below.

config: (jaegerremote.config) {
        sampler: (*jaegerremote.probabilisticSampler)({
            samplingRate: (float64) 0.001,
            samplingBoundary: (uint64) 9223372036854776
        }),
However, sometimes at the starting iterations i still observed below lines where i could see the samplingRate information as "1" which is correct. like below

                                                00000060  6e 67 74 68 3a 20 37 35  0d 0a 0d 0a 7b 22 73 74  |ngth: 75....{"st|
                                                00000070  72 61 74 65 67 79 54 79  70 65 22 3a 22 50 52 4f  |rategyType":"PRO|
                                                00000080  42 41 42 49 4c 49 53 54  49 43 22 2c 22 70 72 6f  |BABILISTIC","pro|
                                                00000090  62 61 62 69 6c 69 73 74  69 63 53 61 6d 70 6c 69  |babilisticSampli|
                                                000000a0  6e 67 22 3a 7b 22 73 61  6d 70 6c 69 6e 67 52 61  |ng":{"samplingRa|
                                                000000b0  74 65 22 3a 31 7d 7d 00  00 00 00 00 00 00 00 00  |te":1}}.


Attached full log here, can you please check if I am doing anything wrong. 
[test.log](https://github.com/jaegertracing/jaeger/files/10427262/test.log)
Please note, I have tested this with TLS disabled, and using curl command able to fetch.
ternminal>$curl 'http://collector_host_name:14268/api/sampling?service=foo'
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":1}}
ternminal>$

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro , Could you please check the logs and help in finding the issue.

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro 

I found the problem why sampling is not working in  without agent case with 14268/api/sampling used, using WithLogger option to checked the logs and I find below error two times with(http://collector_host_name:14268/api/sampling)

And please note with-agent host and 5778/sampling sampling port I did not get below error and sampling is working fine.

_failed to parse sampling strategy response      error: json: cannot unmarshal string into Go struct field SamplingStrategyResponse.strategyType of type api_v2.SamplingStrategyType  
failed to parse sampling strategy response      error: json: cannot unmarshal string into Go struct field SamplingStrategyResponse.strategyType of type api_v2.SamplingStrategyType_

I think this function throwing this error, https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/samplers/jaegerremote/sampler_remote.go#L312

This is my sampling file configured in collector side, 
 "service_strategies": [
             {
              "service": "foo",
              "type": "probabilistic",
              "param": 1.0
             }
          ],
          "default_strategy": {
          "type": "probabilistic",
          "param": 0.8

Could you please suggest what could be issue ?

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro 

Please find few more findings below.Tried three ways from which only one way sampling is working.

**Case1:** With below URL and endpoint, Remote sampling not working, getting bellow error 
**failed to parse sampling strategy response error: json: cannot unmarshal string into Go struct field SamplingStrategyResponse.strategyType of type api_v2.SamplingStrategyType**

terminal>$ curl 'http://collector_host:14268/api/sampling?service=foo'
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":1}}

**Case2:** With below URL and endpoint, Remote sampling not working, getting bellow error 
**failed to parse sampling strategy response error: json: cannot unmarshal string into Go struct field SamplingStrategyResponse.strategyType of type api_v2.SamplingStrategyType**

terminal>$  curl 'http://agent_host:5778/sampling?service=foo'
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":1}}


**Case 3: Below way(without /sampling endpoint) Remote sampling working fine.**
terminal>$ curl 'http://agent_host:5778?service=foo'
{"strategyType":0,"probabilisticSampling":{"samplingRate":1}}

From the above observation, it seems like before sending json parser, strategyType" should be changed 0, instead of passing as "PROBABILISTIC", probably this is causing the issue.

Could you please check and help in solving this issue.

**Reply by yurishkuro** [reply]
Thanks for digging this up, I can see the issue now. It comes from this code:

https://github.com/jaegertracing/jaeger/blob/0237ca4372180cd9640a70e3907d61c03b628f75/pkg/clientcfg/clientcfghttp/handler.go#L175-L185

The back story here is that Jaeger always used Thrift's built-in JSON serialization format. However, up to Thrift v0.9.2 the enums were rendered as integers, and that behavior was changed in Thrift v0.9.3 to render enums as string names. We needed to upgrade Thrift library in Jaeger, but this change was a breaking for existing Jaeger SDKs. The existing SDKs used to query sampling strategies from `agent:5778/` (root endpoint). We've introduced a new endpoint `/sampling` that is now using newer behavior of Thrift serialization (enums as strings), and we changed all Jaeger SDKs to query this new endpoint, but the root `/` endpoint was still kept and was using the old behavior for backwards compatibility.

It would be highly surprising for someone to run into this issue since these changes happened back in 2017 and there should be hardly any SDKs out there that use Thrift v0.9.2  behavior, _including the OpenTelemetry SDK_.

So I checked the linked you provided [1], and that code is not using Thrift at all. Instead, it looks like it's using a Protobuf binding. Jaeger backend does support sampling strategies over gRPC [2], using binary Protobuf messages (that's how Jaeger agent retrieves sampling strategies from Jaeger collector). However, the *JSON format* of sampling strategies is derived fro Thrift representation, not from Protobuf, and OTEL SDK appears to parse that JSON using Protobuf-generated bindings, and therefore fails. I think the fact that this code just happened to work against the old `/` endpoint in the agent is mere a coincidence where Thrift v0.9.2 JSON serialization happened to match Protobuf/JSON.

I assume there is also a related issue in the OTEL Collector that presumably can also serve sampling strategies from files, and I assume it's also using the same deprecated JSON format.

\[1]: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/samplers/jaegerremote/sampler_remote.go#L312
[2]: https://github.com/jaegertracing/jaeger-idl/blob/main/proto/api_v2/sampling.proto#L114

**Reply by Narenderbhcu** [reply]
Hi @yurishkuro , 

Thanks for the answer. we have a question, whether client side(OTEL SDK) code need to be changed or Jaeger side code needs to be changed, in order to fix this issue? And also you mentioned that related issue on OTEL Collector, can you please share that?

**Reply by yurishkuro** [reply]
I will try to do a deep dive this weekend to get a better picture of the state. If OTEL already converged on this (deprecated) format across both OTEL collector and OTEL SDKs then it would make sense to support this format in Jaeger as a new version.

**Reply by Narenderbhcu** [reply]
Okay, we will wait,  thank you so much for the support!

**Reply by yurishkuro** [reply]
I booked a ticket in OTEL: https://github.com/open-telemetry/opentelemetry-specification/issues/3126

**Reply by Narenderbhcu** [reply]
Thanks @yurishkuro, we will monitor.
